# Capstone Projekt Rossmann

# Intro and EDA

# 1. Project definition

## 1.1 Background

This project aims to investigate what insights the drugstore chain Rossmann can gain from its historical sales and advertising data, data on school vacations and public holidays and competitor data. The project will address the question of how this data can be used to optimize store operations and management with the aim of generating more sales. It will also determine how these data sets can be used to predict weekly sales (revenue) for each store with a sufficient level of accuracy. In addition, the possibility of creating automated reports at store level and an overall report is to be created.


Rossmann operates over 4,300 drugstores in 9 European countries. Currently, Rossmann store managers are tasked with preparing their weekly sales forecasts up to eight weeks in advance. Store sales results are influenced by many factors, including promotions, competitive intensity, school and national vacations, seasonal changes and location conditions. Because thousands of individual store managers create sales forecasts based on their individual circumstances, the accuracy of the results can vary widely. Therefore, the company's data science team is on a new mission to create a unified modeling methodology for store managers to predict weekly results with greater accuracy. The management team also needs an overall report with feasible or actionable strategies to understand the overall performance of all stores and find a way to optimize future sales performance (i.e. revenue). Finally, an individual store performance report needs to be provided to each store manager.

## 1.2 General task from the client

- Determine the key factors that influence sales and revenue
- Create strategies for the management to improve the business

- Build a standardized forecasting model which can predict the sales figures for the next eight weeks for each store.

- Generate a report with information on the overall performance of the 1115 stores as well as individual performance reports for each store.

## 1.3 Definition of use cases (background/business purpose)

- The analyses should help to better understand the current sales figures and the impact of e.g. promotions, product range, distance to competitors, etc. on sales.
- The resulting strategy and optimization proposals for increasing sales will then be discussed and implemented at management level.
- The standardized forecast model of sales figures per store for the next eight weeks should also be more accurate than the current individual forecasts of the individual store managers themselves, as well as saving corresponding time, monetary and capacity resources.
- The overall report as well as the individual store reports should also contribute to the mentioned objectives and also provide a quick and automated overview of the current performance.

## 1.4 Formulating hypotheses

1. promotions  
	1.1 In weeks with promotions, sales, number of customers and sales per customer are higher.  
2. promotion2  
	2.1 In months with Promo2, sales, number of customers and sales per customer are higher.  
	2.2 Since the store has been participating in Promo2, sales, number of customers and sales per customer are higher, even outside of Promo2 promotions.  
3. school holidays  
	3.1 Sales are lower in weeks with school vacations.  
	3.2 In weeks with school vacations, sales per customer are lower.  
4. public holidays  
	4.1 Sales are lower in weeks with public holidays.  
	4.2 In weeks with public holidays, more purchases are made on the other days.  
5. day of the week  
	5.1 Fewer purchases are made at the weekend.  
6. assortment  
	6.1 Stores with an extra assortment have higher sales than basic stores  
	6.2 Stores with an extended assortment have higher sales than basic and extra stores  
7. CompetitionDistance  
	7.1 The closer the competitor, the lower the sales.  
	7.2 Sales have been lower since the competitor opened.

## 1.5 Further interesting questions

1. promotions  
	1.1 Is a promotion worthwhile in weeks with public holidays?  
2. promotion2  
3. SchoolHolidays  
4. StateHolidays  
5. day of the week  
	5.1 Is a promotion on the weekend worthwhile?  
6. Assortment  
	6.1 What is the ratio of basic, extra and extended assortment?  
	6.2 What is the ratio of basic, extra and extended assortment in the individual store types?  
	6.3 What is the average turnover of the individual assortments?  
	6.4 How many promotion weeks are there in the individual assortments?  
	6.5 Do promotion weeks perform better in individual assortments than in others?  
7 CompetitionDistance  
8. StoreType  
	8.1 How many stores are there of which type?  
	8.2 How many promotion weeks are there in the individual store types?  
	8.3 Do promotion weeks perform better in individual store types than in others?


# Data Analysis

## Data preparing for EDA/ Row data preparation

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px

import plotly.graph_objects as go
from plotly.subplots import make_subplots

# matplotlib inline notebook





pd.set_option('display.max_columns', None)

### Store Data

In [2]:
df_store = pd.read_csv("store.csv", dtype={0:int, 1:object, 2:object, 3:float, 4:float, 5:float, 6:int, 7:float, 8:float, 9:object})


In [3]:
df_store.sample(10)

Store StoreType Assortment  CompetitionDistance  \
93      94         d          c              14620.0   
250    251         a          c                340.0   
543    544         a          a                250.0   
368    369         d          c               5870.0   
549    550         d          c                 50.0   
713    714         d          c              12070.0   
902    903         d          c               7290.0   
708    709         a          a                500.0   
217    218         a          c                640.0   
238    239         d          c                610.0   

     CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
93                         NaN                       NaN       0   
250                        NaN                       NaN       0   
543                       12.0                    2001.0       1   
368                        4.0                    2014.0       0   
549                        6.0                    2015.0       0   
713                       10.0                    2005.0       1   
902                        9.0                    2014.0       0   
708                       12.0                    2010.0       0   
217                        NaN                       NaN       1   
238                        NaN                       NaN       0   

     Promo2SinceWeek  Promo2SinceYear     PromoInterval  
93               NaN              NaN               NaN  
250              NaN              NaN               NaN  
543             13.0           2010.0   Jan,Apr,Jul,Oct  
368              NaN              NaN               NaN  
549              NaN              NaN               NaN  
713             10.0           2013.0   Jan,Apr,Jul,Oct  
902              NaN              NaN               NaN  
708              NaN              NaN               NaN  
217              9.0           2011.0  Mar,Jun,Sept,Dec  
238              NaN              NaN               NaN

In [4]:
df_store.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1115 entries, 0 to 1114
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Store                      1115 non-null   int32  
 1   StoreType                  1115 non-null   object 
 2   Assortment                 1115 non-null   object 
 3   CompetitionDistance        1112 non-null   float64
 4   CompetitionOpenSinceMonth  761 non-null    float64
 5   CompetitionOpenSinceYear   761 non-null    float64
 6   Promo2                     1115 non-null   int32  
 7   Promo2SinceWeek            571 non-null    float64
 8   Promo2SinceYear            571 non-null    float64
 9   PromoInterval              571 non-null    object 
dtypes: float64(5), int32(2), object(3)
memory usage: 78.5+ KB


In [5]:
for column in df_store.select_dtypes(include=['object']).columns:
    print(f"Unique values in '{column}': {df_store[column].unique()}")

Unique values in 'StoreType': ['c' 'a' 'd' 'b']
Unique values in 'Assortment': ['a' 'c' 'b']
Unique values in 'PromoInterval': [nan 'Jan,Apr,Jul,Oct' 'Feb,May,Aug,Nov' 'Mar,Jun,Sept,Dec']


In [6]:
df_store.describe()

Store  CompetitionDistance  CompetitionOpenSinceMonth  \
count  1115.00000          1112.000000                 761.000000   
mean    558.00000          5404.901079                   7.224704   
std     322.01708          7663.174720                   3.212348   
min       1.00000            20.000000                   1.000000   
25%     279.50000           717.500000                   4.000000   
50%     558.00000          2325.000000                   8.000000   
75%     836.50000          6882.500000                  10.000000   
max    1115.00000         75860.000000                  12.000000   

       CompetitionOpenSinceYear       Promo2  Promo2SinceWeek  Promo2SinceYear  
count                761.000000  1115.000000       571.000000       571.000000  
mean                2008.668857     0.512108        23.595447      2011.763573  
std                    6.195983     0.500078        14.141984         1.674935  
min                 1900.000000     0.000000         1.000000      2009.000000  
25%                 2006.000000     0.000000        13.000000      2011.000000  
50%                 2010.000000     1.000000        22.000000      2012.000000  
75%                 2013.000000     1.000000        37.000000      2013.000000  
max                 2015.000000     1.000000        50.000000      2015.000000

In [ ]:
px.box(df_store.select_dtypes(include=['number']))

### Train Data

In [21]:
df_train = pd.read_csv("train.csv", parse_dates=[2], dtype={7:object})

In [ ]:
# Group by store and aggregate to week
df_train.set_index('Date', inplace=True)
weekly_sales_by_store = df_train.groupby('Store').resample('W').agg({'Sales': 'sum', 'Customers': 'sum', 'Open': 'sum', 'Promo': 'sum', 'StateHoliday': 'sum', 'SchoolHoliday': 'sum'}).reset_index()
weekly_sales_by_store

In [22]:
weekly_sales_by_store = df_train

In [23]:
# Check non mixing state holidays
print("Unique values in 'StateHoliday':", weekly_sales_by_store.StateHoliday.unique())
# Create NumStateHoliday with how many state holidays are in the week
weekly_sales_by_store['StateHoliday'] = weekly_sales_by_store['StateHoliday'].astype(str)
#weekly_sales_by_store['NumStateHoliday'] = weekly_sales_by_store['StateHoliday'].apply(lambda x: x.count('a') + x.count('b') + x.count('c'))
#print("Unique values in 'NumStateHoliday':", weekly_sales_by_store.NumStateHoliday.unique())
# Remove aggregated 0 and doubled labels
#weekly_sales_by_store['StateHoliday'] = weekly_sales_by_store['StateHoliday'].str.extract('(a|b|c)', expand=False).fillna('0')
print("Unique values in 'StateHoliday':", weekly_sales_by_store.StateHoliday.unique())
weekly_sales_by_store

Unique values in 'StateHoliday': ['0' 'a' 'b' 'c']
Unique values in 'StateHoliday': ['0' 'a' 'b' 'c']


Store  DayOfWeek       Date  Sales  Customers  Open  Promo  \
0            1          5 2015-07-31   5263        555     1      1   
1            2          5 2015-07-31   6064        625     1      1   
2            3          5 2015-07-31   8314        821     1      1   
3            4          5 2015-07-31  13995       1498     1      1   
4            5          5 2015-07-31   4822        559     1      1   
...        ...        ...        ...    ...        ...   ...    ...   
1017204   1111          2 2013-01-01      0          0     0      0   
1017205   1112          2 2013-01-01      0          0     0      0   
1017206   1113          2 2013-01-01      0          0     0      0   
1017207   1114          2 2013-01-01      0          0     0      0   
1017208   1115          2 2013-01-01      0          0     0      0   

        StateHoliday  SchoolHoliday  
0                  0              1  
1                  0              1  
2                  0              1  
3                  0              1  
4                  0              1  
...              ...            ...  
1017204            a              1  
1017205            a              1  
1017206            a              1  
1017207            a              1  
1017208            a              1  

[1017209 rows x 9 columns]

In [24]:
weekly_sales_by_store[weekly_sales_by_store['StateHoliday'] != '0']

Store  DayOfWeek       Date  Sales  Customers  Open  Promo  \
63555        1          4 2015-06-04      0          0     0      1   
63557        3          4 2015-06-04      0          0     0      1   
63559        5          4 2015-06-04   5807        625     1      1   
63560        6          4 2015-06-04   5749        614     1      1   
63563        9          4 2015-06-04      0          0     0      1   
...        ...        ...        ...    ...        ...   ...    ...   
1017204   1111          2 2013-01-01      0          0     0      0   
1017205   1112          2 2013-01-01      0          0     0      0   
1017206   1113          2 2013-01-01      0          0     0      0   
1017207   1114          2 2013-01-01      0          0     0      0   
1017208   1115          2 2013-01-01      0          0     0      0   

        StateHoliday  SchoolHoliday  
63555              a              0  
63557              a              0  
63559              a              0  
63560              a              0  
63563              a              0  
...              ...            ...  
1017204            a              1  
1017205            a              1  
1017206            a              1  
1017207            a              1  
1017208            a              1  

[31050 rows x 9 columns]

In [25]:
# Create Month and Year columns
#weekly_sales_by_store.insert(weekly_sales_by_store.columns.get_loc('Date') + 1, 'CW', weekly_sales_by_store['Date'].dt.isocalendar().week)
weekly_sales_by_store.insert(weekly_sales_by_store.columns.get_loc('Date') + 2, 'Month', weekly_sales_by_store['Date'].dt.month)
#weekly_sales_by_store.insert(weekly_sales_by_store.columns.get_loc('Date') + 3, 'Year', weekly_sales_by_store['Date'].dt.year)
#weekly_sales_by_store.insert(weekly_sales_by_store.columns.get_loc('Date') + 4, 'DayOfWeek', weekly_sales_by_store['Date'].dt.dayofweek)

# Create SalesPerCustomer column
weekly_sales_by_store.insert(weekly_sales_by_store.columns.get_loc('Sales') + 1, 'SalesPerCustomer', weekly_sales_by_store['Sales'] / weekly_sales_by_store['Customers'])
# Create sales per open day
weekly_sales_by_store.insert(weekly_sales_by_store.columns.get_loc('Sales') + 2, 'SalesPerOpenDay', weekly_sales_by_store['Sales'] / weekly_sales_by_store['Open'])
# Create customers per open day
weekly_sales_by_store.insert(weekly_sales_by_store.columns.get_loc('Customers') + 1, 'CustomersPerOpenDay', weekly_sales_by_store['Customers'] / weekly_sales_by_store['Open'])

weekly_sales_by_store.insert(weekly_sales_by_store.columns.get_loc('Promo') + 1, 'IsPromo', np.where(weekly_sales_by_store['Promo'] > 0, 1, 0))
weekly_sales_by_store.insert(weekly_sales_by_store.columns.get_loc('StateHoliday') + 1, 'IsStateHoliday', np.where(weekly_sales_by_store['StateHoliday'] != '0', 1, 0))
#weekly_sales_by_store.insert(weekly_sales_by_store.columns.get_loc('SchoolHoliday') + 1, 'IsSchoolHoliday', np.where(weekly_sales_by_store['SchoolHoliday'] > 0, 1, 0))
weekly_sales_by_store


Store  DayOfWeek       Date  Sales  SalesPerCustomer  \
0            1          5 2015-07-31   5263          9.482883   
1            2          5 2015-07-31   6064          9.702400   
2            3          5 2015-07-31   8314         10.126675   
3            4          5 2015-07-31  13995          9.342457   
4            5          5 2015-07-31   4822          8.626118   
...        ...        ...        ...    ...               ...   
1017204   1111          2 2013-01-01      0               NaN   
1017205   1112          2 2013-01-01      0               NaN   
1017206   1113          2 2013-01-01      0               NaN   
1017207   1114          2 2013-01-01      0               NaN   
1017208   1115          2 2013-01-01      0               NaN   

         SalesPerOpenDay  Month  Customers  CustomersPerOpenDay  Open  Promo  \
0                 5263.0      7        555                555.0     1      1   
1                 6064.0      7        625                625.0     1      1   
2                 8314.0      7        821                821.0     1      1   
3                13995.0      7       1498               1498.0     1      1   
4                 4822.0      7        559                559.0     1      1   
...                  ...    ...        ...                  ...   ...    ...   
1017204              NaN      1          0                  NaN     0      0   
1017205              NaN      1          0                  NaN     0      0   
1017206              NaN      1          0                  NaN     0      0   
1017207              NaN      1          0                  NaN     0      0   
1017208              NaN      1          0                  NaN     0      0   

         IsPromo StateHoliday  IsStateHoliday  SchoolHoliday  
0              1            0               0              1  
1              1            0               0              1  
2              1            0               0              1  
3              1            0               0              1  
4              1            0               0              1  
...          ...          ...             ...            ...  
1017204        0            a               1              1  
1017205        0            a               1              1  
1017206        0            a               1              1  
1017207        0            a               1              1  
1017208        0            a               1              1  

[1017209 rows x 15 columns]

In [26]:
# Merge df_train with df_store
weekly_sales_with_store_info = pd.merge(weekly_sales_by_store, df_store, on='Store', how='left')
# Check if all sores were found
print("Amount of none found stores:", weekly_sales_with_store_info.Date.isna().sum())

Amount of none found stores: 0


In [27]:
# Create Promo2Member column too see if the store joined the Promo2 program
def is_promo2_active(row):
    if row['Promo2'] == 1:
        # Construct the starting date of Promo2
        promo2_start_date = datetime.fromisocalendar(int(row['Promo2SinceYear']), int(row['Promo2SinceWeek']), 1)
        # Check if the promo was active on the 'Date'
        return 1 if promo2_start_date <= row['Date'] else 0
    else:
        # If Promo2 is not active, return False
        return 0

# Apply the function across the DataFrame rows
weekly_sales_with_store_info['Promo2Member'] = weekly_sales_with_store_info.apply(is_promo2_active, axis=1)

In [28]:
# Create Promo2Active if current month is in PromoInterval
month_dict = {1: 'Jan', 2: 'Feb', 3: 'Mar', 4: 'Apr', 5: 'May', 6: 'Jun', 7: 'Jul', 8: 'Aug', 9: 'Sep', 10: 'Oct', 11: 'Nov', 12: 'Dec'}
# if Promo2Member == 1 and Month(conver with month_dict) is in PromoIntervall(has to be splitted), then Promo2Active = 1
def is_promo2_active_month(row):
	if row['Promo2Member'] == 1:
		# Check if the promo was active on the 'Date'
		return 1 if month_dict[row['Month']] in row['PromoInterval'].split(',') else 0
	else:
		# If Promo2 is not active, return False
		return 0

# Apply the function across the DataFrame rows
weekly_sales_with_store_info['Promo2Active'] = weekly_sales_with_store_info.apply(is_promo2_active_month, axis=1)


In [29]:
# Create IsCompetition column

def is_competition_active(row):
	if row['CompetitionOpenSinceYear'] > 0:
		# Construct the starting date of competition
		competition_start_date = datetime(int(row['CompetitionOpenSinceYear']), int(row['CompetitionOpenSinceMonth']), 1)
		# Check if the competition was active on the 'Date'
		return 1 if competition_start_date <= row['Date'] else 0
	else:
		# If competition is not active, return False
		return 0

# Apply the function across the DataFrame rows
weekly_sales_with_store_info.insert(weekly_sales_with_store_info.columns.get_loc('CompetitionOpenSinceYear') + 1, 'IsCompetition', weekly_sales_with_store_info.apply(is_competition_active, axis=1))
#weekly_sales_with_store_info.insert(weekly_sales_with_store_info.columns.get_loc('CompetitionOpenSinceYear') + 1, 'IsCompetition', np.where(weekly_sales_with_store_info['CompetitionOpenSinceYear'] > 0, 1, 0))

In [30]:
print(weekly_sales_with_store_info.info())
weekly_sales_with_store_info.sample(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 27 columns):
 #   Column                     Non-Null Count    Dtype         
---  ------                     --------------    -----         
 0   Store                      1017209 non-null  int64         
 1   DayOfWeek                  1017209 non-null  int64         
 2   Date                       1017209 non-null  datetime64[ns]
 3   Sales                      1017209 non-null  int64         
 4   SalesPerCustomer           844340 non-null   float64       
 5   SalesPerOpenDay            844392 non-null   float64       
 6   Month                      1017209 non-null  int32         
 7   Customers                  1017209 non-null  int64         
 8   CustomersPerOpenDay        844392 non-null   float64       
 9   Open                       1017209 non-null  int64         
 10  Promo                      1017209 non-null  int64         
 11  IsPromo                    1017209 no

Store  DayOfWeek       Date  Sales  SalesPerCustomer  SalesPerOpenDay  \
375294    641          2 2014-08-05   7177         11.862810           7177.0   
997096   1072          6 2013-01-19   3773         10.197297           3773.0   
339746    616          5 2014-09-12   7145          9.090331           7145.0   
556353    754          2 2014-02-18   6738          7.753740           6738.0   
951436     12          4 2013-02-28   6210          7.704715           6210.0   

        Month  Customers  CustomersPerOpenDay  Open  Promo  IsPromo  \
375294      8        605                605.0     1      1        1   
997096      1        370                370.0     1      0        0   
339746      9        786                786.0     1      0        0   
556353      2        869                869.0     1      1        1   
951436      2        806                806.0     1      0        0   

       StateHoliday  IsStateHoliday  SchoolHoliday StoreType Assortment  \
375294            0               0              1         a          c   
997096            0               0              0         a          c   
339746            0               0              0         a          c   
556353            0               0              1         c          c   
951436            0               0              0         a          c   

        CompetitionDistance  CompetitionOpenSinceMonth  \
375294                610.0                       12.0   
997096               5380.0                        8.0   
339746               3040.0                        8.0   
556353                380.0                        5.0   
951436               1070.0                        NaN   

        CompetitionOpenSinceYear  IsCompetition  Promo2  Promo2SinceWeek  \
375294                    2003.0              1       1             36.0   
997096                    2015.0              0       1              5.0   
339746                    2011.0              1       1             35.0   
556353                    2008.0              1       1             10.0   
951436                       NaN              0       1             13.0   

        Promo2SinceYear     PromoInterval  Promo2Member  Promo2Active  
375294           2013.0  Mar,Jun,Sept,Dec             1             0  
997096           2010.0   Feb,May,Aug,Nov             1             0  
339746           2010.0  Mar,Jun,Sept,Dec             1             0  
556353           2014.0  Mar,Jun,Sept,Dec             0             0  
951436           2010.0   Jan,Apr,Jul,Oct             1             0

In [ ]:
used_df = weekly_sales_with_store_info[['StoreType', 'Sales', 'Customers']]
px.box(used_df, color='StoreType')

In [32]:
# As the store were closed, we can fill the nans with 0

# fill nans with 0 for listed columns
columns_to_fill = ['SalesPerCustomer', 'SalesPerOpenDay', 'CustomersPerOpenDay']
weekly_sales_with_store_info = weekly_sales_with_store_info.fillna({col: 0 for col in columns_to_fill})
# As if the store is not participating in Promo2, PromoInterval is 0, we can fill the nans with 0
weekly_sales_with_store_info['PromoInterval'] = weekly_sales_with_store_info['PromoInterval'].fillna(0)
weekly_sales_with_store_info

Store  DayOfWeek       Date  Sales  SalesPerCustomer  \
0            1          5 2015-07-31   5263          9.482883   
1            2          5 2015-07-31   6064          9.702400   
2            3          5 2015-07-31   8314         10.126675   
3            4          5 2015-07-31  13995          9.342457   
4            5          5 2015-07-31   4822          8.626118   
...        ...        ...        ...    ...               ...   
1017204   1111          2 2013-01-01      0          0.000000   
1017205   1112          2 2013-01-01      0          0.000000   
1017206   1113          2 2013-01-01      0          0.000000   
1017207   1114          2 2013-01-01      0          0.000000   
1017208   1115          2 2013-01-01      0          0.000000   

         SalesPerOpenDay  Month  Customers  CustomersPerOpenDay  Open  Promo  \
0                 5263.0      7        555                555.0     1      1   
1                 6064.0      7        625                625.0     1      1   
2                 8314.0      7        821                821.0     1      1   
3                13995.0      7       1498               1498.0     1      1   
4                 4822.0      7        559                559.0     1      1   
...                  ...    ...        ...                  ...   ...    ...   
1017204              0.0      1          0                  0.0     0      0   
1017205              0.0      1          0                  0.0     0      0   
1017206              0.0      1          0                  0.0     0      0   
1017207              0.0      1          0                  0.0     0      0   
1017208              0.0      1          0                  0.0     0      0   

         IsPromo StateHoliday  IsStateHoliday  SchoolHoliday StoreType  \
0              1            0               0              1         c   
1              1            0               0              1         a   
2              1            0               0              1         a   
3              1            0               0              1         c   
4              1            0               0              1         a   
...          ...          ...             ...            ...       ...   
1017204        0            a               1              1         a   
1017205        0            a               1              1         c   
1017206        0            a               1              1         a   
1017207        0            a               1              1         a   
1017208        0            a               1              1         d   

        Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0                a               1270.0                        9.0   
1                a                570.0                       11.0   
2                a              14130.0                       12.0   
3                c                620.0                        9.0   
4                a              29910.0                        4.0   
...            ...                  ...                        ...   
1017204          a               1900.0                        6.0   
1017205          c               1880.0                        4.0   
1017206          c               9260.0                        NaN   
1017207          c                870.0                        NaN   
1017208          c               5350.0                        NaN   

         CompetitionOpenSinceYear  IsCompetition  Promo2  Promo2SinceWeek  \
0                          2008.0              1       0              NaN   
1                          2007.0              1       1             13.0   
2                          2006.0              1       1             14.0   
3                          2009.0              1       0              NaN   
4                          2015.0              1       0              NaN   
...                           ...            ...     ...              ...  

In [39]:
# import typesrom pandas.api.types import infer_dtype
from pandas.api.types import infer_dtype

# check if a column contains mixed data typesrom pandas.api.types import infer_dtype

#for col in ['StateHoliday', 'StoreType', 'Assortment', 'PromoInterval']:
for col in weekly_sales_with_store_info.columns:
    dtype = infer_dtype(weekly_sales_with_store_info[col])
    print(f"Data type of {col}: {dtype}")
    
# Convert mixed columns
cols_to_convert = ['PromoInterval']
weekly_sales_with_store_info[cols_to_convert] = weekly_sales_with_store_info[cols_to_convert].astype(str)

Data type of Store: integer
Data type of DayOfWeek: integer
Data type of Date: datetime64
Data type of Sales: integer
Data type of SalesPerCustomer: floating
Data type of SalesPerOpenDay: floating
Data type of Month: integer
Data type of Customers: integer
Data type of CustomersPerOpenDay: floating
Data type of Open: integer
Data type of Promo: integer
Data type of IsPromo: integer
Data type of StateHoliday: string
Data type of IsStateHoliday: integer
Data type of SchoolHoliday: integer
Data type of StoreType: string
Data type of Assortment: string
Data type of CompetitionDistance: floating
Data type of CompetitionOpenSinceMonth: floating
Data type of CompetitionOpenSinceYear: floating
Data type of IsCompetition: integer
Data type of Promo2: integer
Data type of Promo2SinceWeek: floating
Data type of Promo2SinceYear: floating
Data type of PromoInterval: string
Data type of Promo2Member: integer
Data type of Promo2Active: integer


In [38]:
#save df to csv
weekly_sales_with_store_info.to_csv("daily_sales_with_store_info.csv", index=False)


# EDA

#### Test Prophet with daily data

In [43]:
import pandas as pd
import numpy as np
from datetime import datetime

from pandas.api.types import infer_dtype

from prophet import Prophet

from sklearn.metrics import mean_absolute_error as mae, mean_squared_error as mse, r2_score
from sklearn.metrics import mean_absolute_error, r2_score
from math import sqrt

In [44]:
df = pd.read_csv('daily_sales_with_store_info.csv')
df['Date'] = pd.to_datetime(df['Date'])
df

C:\Users\Chris\AppData\Local\Temp\ipykernel_13900\2005776927.py:1: DtypeWarning: Columns (12) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('daily_sales_with_store_info.csv')


Store  DayOfWeek       Date  Sales  SalesPerCustomer  \
0            1          5 2015-07-31   5263          9.482883   
1            2          5 2015-07-31   6064          9.702400   
2            3          5 2015-07-31   8314         10.126675   
3            4          5 2015-07-31  13995          9.342457   
4            5          5 2015-07-31   4822          8.626118   
...        ...        ...        ...    ...               ...   
1017204   1111          2 2013-01-01      0          0.000000   
1017205   1112          2 2013-01-01      0          0.000000   
1017206   1113          2 2013-01-01      0          0.000000   
1017207   1114          2 2013-01-01      0          0.000000   
1017208   1115          2 2013-01-01      0          0.000000   

         SalesPerOpenDay  Month  Customers  CustomersPerOpenDay  Open  Promo  \
0                 5263.0      7        555                555.0     1      1   
1                 6064.0      7        625                625.0     1      1   
2                 8314.0      7        821                821.0     1      1   
3                13995.0      7       1498               1498.0     1      1   
4                 4822.0      7        559                559.0     1      1   
...                  ...    ...        ...                  ...   ...    ...   
1017204              0.0      1          0                  0.0     0      0   
1017205              0.0      1          0                  0.0     0      0   
1017206              0.0      1          0                  0.0     0      0   
1017207              0.0      1          0                  0.0     0      0   
1017208              0.0      1          0                  0.0     0      0   

         IsPromo StateHoliday  IsStateHoliday  SchoolHoliday StoreType  \
0              1            0               0              1         c   
1              1            0               0              1         a   
2              1            0               0              1         a   
3              1            0               0              1         c   
4              1            0               0              1         a   
...          ...          ...             ...            ...       ...   
1017204        0            a               1              1         a   
1017205        0            a               1              1         c   
1017206        0            a               1              1         a   
1017207        0            a               1              1         a   
1017208        0            a               1              1         d   

        Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0                a               1270.0                        9.0   
1                a                570.0                       11.0   
2                a              14130.0                       12.0   
3                c                620.0                        9.0   
4                a              29910.0                        4.0   
...            ...                  ...                        ...   
1017204          a               1900.0                        6.0   
1017205          c               1880.0                        4.0   
1017206          c               9260.0                        NaN   
1017207          c                870.0                        NaN   
1017208          c               5350.0                        NaN   

         CompetitionOpenSinceYear  IsCompetition  Promo2  Promo2SinceWeek  \
0                          2008.0              1       0              NaN   
1                          2007.0              1       1             13.0   
2                          2006.0              1       1             14.0   
3                          2009.0              1       0              NaN   
4                          2015.0              1       0              NaN   
...                           ...            ...     ...              ...  

In [45]:


# Daten laden und vorbereiten
df = pd.read_csv('daily_sales_with_store_info.csv', low_memory=False)
df['Date'] = pd.to_datetime(df['Date'])
df = df[['Store', 'Date', 'Sales', 'IsPromo', 'IsStateHoliday', 'Open', 'SchoolHoliday']]

# Konstanten definieren
amount_test_days = 8 * 7  # 8 Wochen * 7 Tage
rolling_mean_days = 4 * 7  # 4 Wochen * 7 Tage

# Ergebnisarrays initialisieren
results_R2_model_train = []
results_MAE_model_train = []
results_R2_model_test = []
results_MAE_model_test = []
results_R2_mean = []
results_MAE_mean = []

for store_id in df['Store'].unique():
    df_store = df[df['Store'] == store_id].sort_values(by='Date')
    
    # prepare holiday information
    listOfStateholidays = df_store[df_store['IsStateHoliday'] == 1]['Date'].to_list()
    stateHolidays = pd.DataFrame({'holiday': 'stateHoliday', 'ds': listOfStateholidays})
    listOfSchoolholidays = df_store[df_store['SchoolHoliday'] == 1]['Date'].to_list()
    schoolHolidays = pd.DataFrame({'holiday': 'schoolHoliday', 'ds': listOfSchoolholidays})
    holidays_df = pd.concat((stateHolidays, schoolHolidays))
    
    df_store = df_store.rename(columns={'Date': 'ds', 'Sales': 'y'}).drop(columns=['Store'])
    
    train_df = df_store[:-amount_test_days].copy()
    test_df = df_store[-amount_test_days:].copy()
    
    # Prophet-Modell erstellen und anpassen
    model = Prophet(holidays=holidays_df)
    model.add_regressor('IsPromo')
    #model.add_regressor('IsStateHoliday')
    #model.add_regressor('SchoolHoliday')
    model.add_regressor('Open')
    model.fit(train_df)
    
    # Vorhersagen und Leistungsbewertung für das Testset
    forecast_test = model.predict(test_df.drop(columns=['y']))
    test_df['week'] = test_df['ds'].dt.strftime('%Y-%U')
    forecast_test['week'] = forecast_test['ds'].dt.strftime('%Y-%U')
    
    weekly_actual_test = test_df.groupby('week')['y'].sum()
    weekly_pred_test = forecast_test.groupby('week')['yhat'].sum()
    
    results_MAE_model_test.append(mean_absolute_error(weekly_actual_test, weekly_pred_test))
    results_R2_model_test.append(r2_score(weekly_actual_test, weekly_pred_test))
    
    # Vorhersagen und Leistungsbewertung für das Trainingset
    forecast_train = model.predict(train_df.drop(columns=['y']))
    train_df['week'] = train_df['ds'].dt.strftime('%Y-%U')
    forecast_train['week'] = forecast_train['ds'].dt.strftime('%Y-%U')
    
    weekly_actual_train = train_df.groupby('week')['y'].sum()
    weekly_pred_train = forecast_train.groupby('week')['yhat'].sum()
    
    results_MAE_model_train.append(mean_absolute_error(weekly_actual_train, weekly_pred_train))
    results_R2_model_train.append(r2_score(weekly_actual_train, weekly_pred_train))
    
    # Berechnung der rollierenden Mittelwerte für das Testdatenset
    rolling_means = train_df['y'].rolling(window=rolling_mean_days, min_periods=1).mean().iloc[-amount_test_days:].groupby(train_df['week']).mean().reset_index(drop=True)
    rolling_weekly_actual_test = weekly_actual_test.reset_index(drop=True)
    results_MAE_mean.append(mean_absolute_error(rolling_weekly_actual_test, rolling_means))
    results_R2_mean.append(r2_score(rolling_weekly_actual_test, rolling_means))

# Durchschnittsergebnisse für das Modell und die rollierenden Mittelwerte
print("Prophet Modell Training: MAE:", np.mean(results_MAE_model_train), "R2:", np.mean(results_R2_model_train))
print("Prophet Modell Test: MAE:", np.mean(results_MAE_model_test), "R2:", np.mean(results_R2_model_test))
print("Rolling Mean: MAE:", np.mean(results_MAE_mean), "R2:", np.mean(results_R2_mean))


17:13:24 - cmdstanpy - INFO - Chain [1] start processing
17:13:24 - cmdstanpy - INFO - Chain [1] done processing
17:13:25 - cmdstanpy - INFO - Chain [1] start processing
17:13:25 - cmdstanpy - INFO - Chain [1] done processing
17:13:26 - cmdstanpy - INFO - Chain [1] start processing
17:13:26 - cmdstanpy - INFO - Chain [1] done processing
17:13:26 - cmdstanpy - INFO - Chain [1] start processing
17:13:26 - cmdstanpy - INFO - Chain [1] done processing
17:13:27 - cmdstanpy - INFO - Chain [1] start processing
17:13:27 - cmdstanpy - INFO - Chain [1] done processing
17:13:28 - cmdstanpy - INFO - Chain [1] start processing
17:13:28 - cmdstanpy - INFO - Chain [1] done processing
17:13:28 - cmdstanpy - INFO - Chain [1] start processing
17:13:28 - cmdstanpy - INFO - Chain [1] done processing
17:13:29 - cmdstanpy - INFO - Chain [1] start processing
17:13:29 - cmdstanpy - INFO - Chain [1] done processing
17:13:30 - cmdstanpy - INFO - Chain [1] start processing
17:13:30 - cmdstanpy - INFO - Chain [1]

Prophet Modell Training: MAE: 2314.915876654955 R2: 0.8828246818493316
Prophet Modell Test: MAE: 2437.331815756532 R2: 0.9304597343706456
Rolling Mean: MAE: 31661.088129810967 R2: -6.2373662833483365


**Result:**
--> Just 43 MAE better then the weekly model

Prophet Modell Training: MAE: 2314.915876654955 R2: 0.8828246818493316
Prophet Modell Test: MAE: 2437.331815756532 R2: 0.9304597343706456
Rolling Mean: MAE: 31661.088129810967 R2: -6.2373662833483365